File: config/tmux/session-manager/tracker.js
/**
 * Session Access Tracker
 * Manages persistent state for session access history
 */

import fs from "fs";
import path from "path";
import os from "os";

// State file location
const STATE_DIR = path.join(os.homedir(), ".local", "share", "tmux");
const STATE_FILE = path.join(STATE_DIR, "session-history.json");

/**
 * Ensures the state directory exists
 */
function ensureStateDirectory() {
  if (!fs.existsSync(STATE_DIR)) {
    fs.mkdirSync(STATE_DIR, { recursive: true });
  }
}

/**
 * Loads session access history from disk
 * @returns {Object} Session history data
 */
function loadSessionHistory() {
  try {
    if (!fs.existsSync(STATE_FILE)) {
      return { sessions: {}, lastAccessed: [] };
    }

    const data = fs.readFileSync(STATE_FILE, "utf8");
    const parsed = JSON.parse(data);

    // Validate structure
    return {
      sessions: parsed.sessions || {},
      lastAccessed: parsed.lastAccessed || [],
    };
  } catch (error) {
    console.error(`Warning: Could not load session history: ${error.message}`);
    return { sessions: {}, lastAccessed: [] };
  }
}

/**
 * Saves session access history to disk
 * @param {Object} history - Session history data
 */
function saveSessionHistory(history) {
  try {
    ensureStateDirectory();
    fs.writeFileSync(STATE_FILE, JSON.stringify(history, null, 2));
  } catch (error) {
    console.error(`Warning: Could not save session history: ${error.message}`);
  }
}

/**
 * Records access to a session
 * @param {string} sessionName - Name of the session
 */
export function recordSessionAccess(sessionName) {
  const history = loadSessionHistory();
  const now = Date.now();

  // Update session access data
  if (!history.sessions[sessionName]) {
    history.sessions[sessionName] = {
      accessCount: 0,
      firstAccessed: now,
      lastAccessed: now,
    };
  }

  history.sessions[sessionName].accessCount++;
  history.sessions[sessionName].lastAccessed = now;

  // Update recent access list (keep only last 20)
  history.lastAccessed = history.lastAccessed.filter((name) => name !== sessionName);
  history.lastAccessed.unshift(sessionName);
  history.lastAccessed = history.lastAccessed.slice(0, 20);

  saveSessionHistory(history);
}

/**
 * Gets session access data
 * @param {string} sessionName - Name of the session
 * @returns {Object|null} Session access data or null if not found
 */
export function getSessionAccess(sessionName) {
  const history = loadSessionHistory();
  return history.sessions[sessionName] || null;
}

/**
 * Gets recently accessed sessions in order
 * @returns {string[]} Array of session names in recent access order
 */
export function getRecentlyAccessedSessions() {
  const history = loadSessionHistory();
  return history.lastAccessed;
}

/**
 * Gets all session access data
 * @returns {Object} All session access data
 */
export function getAllSessionHistory() {
  return loadSessionHistory();
}

/**
 * Cleans up old session data (removes sessions not accessed in 30 days)
 */
export function cleanupOldSessions() {
  const history = loadSessionHistory();
  const thirtyDaysAgo = Date.now() - 30 * 24 * 60 * 60 * 1000;

  let hasChanges = false;

  // Remove old sessions
  for (const [sessionName, data] of Object.entries(history.sessions)) {
    if (data.lastAccessed < thirtyDaysAgo) {
      delete history.sessions[sessionName];
      hasChanges = true;
    }
  }

  // Clean up recent access list
  const oldLength = history.lastAccessed.length;
  history.lastAccessed = history.lastAccessed.filter(
    (sessionName) => history.sessions[sessionName]
  );

  if (history.lastAccessed.length !== oldLength) {
    hasChanges = true;
  }

  if (hasChanges) {
    saveSessionHistory(history);
  }
}

// CLI functionality
if (process.argv[1] === new URL(import.meta.url).pathname) {
  const command = process.argv[2];

  switch (command) {
    case "record":
      const sessionName = process.argv[3];
      if (sessionName) {
        recordSessionAccess(sessionName);
        console.log(`Recorded access to session: ${sessionName}`);
      } else {
        console.error("Usage: session-tracker.js record <session-name>");
      }
      break;

    case "list":
      const recent = getRecentlyAccessedSessions();
      console.log("Recently accessed sessions:");
      recent.forEach((name, index) => {
        console.log(`  ${index + 1}. ${name}`);
      });
      break;

    case "cleanup":
      cleanupOldSessions();
      console.log("Cleaned up old session data");
      break;

    case "show":
      const allHistory = getAllSessionHistory();
      console.log(JSON.stringify(allHistory, null, 2));
      break;

    default:
      console.log("Usage: session-tracker.js <command>");
      console.log("Commands:");
      console.log("  record <session-name>  - Record access to a session");
      console.log("  list                   - List recently accessed sessions");
      console.log("  cleanup                - Clean up old session data");
      console.log("  show                   - Show all session history");
  }
}

File: packages/tsm/tracker.go
package main

import (
	"encoding/json"
	"os"
	"path/filepath"
	"time"
)

const maxRecentSessions = 20
const staleSessionDays = 30

type sessionEntry struct {
	AccessCount   int   `json:"accessCount"`
	FirstAccessed int64 `json:"firstAccessed"`
	LastAccessed  int64 `json:"lastAccessed"`
}

type sessionHistory struct {
	Sessions     map[string]sessionEntry `json:"sessions"`
	LastAccessed []string                `json:"lastAccessed"`
}

func historyFile() string {
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".local", "share", "tmux", "session-history.json")
}

func loadHistory() sessionHistory {
	h := sessionHistory{Sessions: make(map[string]sessionEntry)}
	data, err := os.ReadFile(historyFile())
	if err != nil {
		return h
	}
	_ = json.Unmarshal(data, &h)
	if h.Sessions == nil {
		h.Sessions = make(map[string]sessionEntry)
	}
	return h
}

func saveHistory(h sessionHistory) {
	f := historyFile()
	_ = os.MkdirAll(filepath.Dir(f), 0o755)
	data, _ := json.MarshalIndent(h, "", "  ")
	_ = os.WriteFile(f, data, 0o644)
}

// RecordSessionAccess bumps the access count and prepends the session
// to the recently-accessed list (capped at maxRecentSessions).
func RecordSessionAccess(name string) {
	h := loadHistory()
	now := time.Now().UnixMilli()

	entry := h.Sessions[name]
	if entry.FirstAccessed == 0 {
		entry.FirstAccessed = now
	}
	entry.AccessCount++
	entry.LastAccessed = now
	h.Sessions[name] = entry

	filtered := make([]string, 0, len(h.LastAccessed))
	for _, n := range h.LastAccessed {
		if n != name {
			filtered = append(filtered, n)
		}
	}
	h.LastAccessed = append([]string{name}, filtered...)
	if len(h.LastAccessed) > maxRecentSessions {
		h.LastAccessed = h.LastAccessed[:maxRecentSessions]
	}

	saveHistory(h)
}

// RecentSessions returns the recently-accessed session names in order.
func RecentSessions() []string {
	return loadHistory().LastAccessed
}

// CleanupOldSessions removes sessions not accessed in staleSessionDays days.
func CleanupOldSessions() {
	h := loadHistory()
	cutoff := time.Now().AddDate(0, 0, -staleSessionDays).UnixMilli()

	changed := false
	for name, entry := range h.Sessions {
		if entry.LastAccessed < cutoff {
			delete(h.Sessions, name)
			changed = true
		}
	}

	filtered := make([]string, 0, len(h.LastAccessed))
	for _, name := range h.LastAccessed {
		if _, ok := h.Sessions[name]; ok {
			filtered = append(filtered, name)
		}
	}
	if len(filtered) != len(h.LastAccessed) {
		h.LastAccessed = filtered
		changed = true
	}

	if changed {
		saveHistory(h)
	}
}

// AllHistory returns the raw history struct (for the `tracker show` command).
func AllHistory() sessionHistory {
	return loadHistory()
}
